### PR TITLE
fix(avalonia): Map Action Animation expand repoints LDR-literal references (#1025)

### LINKED ISSUE #1025: Avalonia Map Action Animation — list expansion does not rescan LDR literal-pool pointers
## Summary
The Map Action Animation list-expand repoints only the primary table pointer; references reached via ARM Thumb LDR literal-pool loads are not rescanned (KnownGap #501). The opt-in Core `RepointAllReferences` helper exists but is not used here.

## Current behaviour (evidence)
- `FEBuilderGBA.Avalonia/ViewModels/ImageMapActionAnimationViewModel.cs:505-509` — `ExpandList` XML doc states it mirrors WF expand "but without the LDR-pointer rescan (KnownGap per #501)."
- `ImageMapActionAnimationViewModel.cs:537` — `ExpandList` delegates to `DataExpansionCore.ExpandTableTo`, which does a single-slot repoint only (`rom.write_p32(pointerAddr, newBase)` at `FEBuilderGBA.Core/DataExpansionCore.cs:397`).
- `FEBuilderGBA.Core/DataExpansionCore.cs:472` — the opt-in `RepointAllReferences` helper (which calls `U.GrepPointerAllOnLDR` to repoint LDR-literal references) exists but is NOT invoked by `ExpandList`.

## Expected / WinForms reference
WF `InputFormRef.ExpandsArea` + `MoveToFreeSapceForm.SearchPointer` use `GrepPointerAllOnLDR` (present in `FEBuilderGBA/MoveToFreeSapceForm.cs`) to repoint all references including LDR literal pools.

## Acceptance criteria
- Optionally route the expand through `DataExpansionCore.RepointAllReferences` to also repoint LDR-literal references.
- Edge-case parity gap; low priority — the expand button works and repoints the primary pointer.

---
<sub>🔎 Filed by [Claude Code](https://claude.com/claude-code) `v2.1.168` · model `claude-opus-4-8[1m]` — part of an automated **Avalonia GUI stub audit** (66-agent workflow: 132 candidate files deep-scanned + whole-tree semantic sweep → 192 stubs found → 46 grouped → 43 adversarially verified + 1 critic-found).</sub>

### PR DESCRIPTION
**Pull request by laqieer**
## Summary
Closes the KnownGap (#501) where the Map Action Animation list-expand repointed **only** the primary table pointer. `ImageMapActionAnimationViewModel.ExpandList` now threads the active `UndoData` and, after `DataExpansionCore.ExpandTableTo` relocates the table, calls the existing tested helper `DataExpansionCore.RepointAllReferences` (#781) to repoint **every** raw 32-bit pointer **and** ARM-Thumb **LDR literal-pool** reference to the old base. This mirrors the already-merged `WorldMapImageViewModel` / `EventMapChangeViewModel` expand pattern. **No Core changes.**

`RepointAllReferences` is the **last ROM-mutating step** (the canonical slot is already moved by `ExpandTableTo`, so it isn't double-written; a `0` return = no secondary refs = success, never a rollback).

Plan + Copilot CLI review: issue #1025 comment thread (no correctness blocker; approval conditioned on the behavior-level LDR+undo test — included).

## Scope
Closes #1025.

## Test plan
- [x] `dotnet build FEBuilderGBA.Avalonia` — Release, 0 errors
- [x] `FEBuilderGBA.Avalonia.Tests` — **7638/7641** pass (3 skipped, **0 failures**); filtered `~ImageMapActionAnimation` **31/31**
- [x] `FEBuilderGBA.Tests` (net9.0-windows) — **1851/1851** pass
- [x] `ViewModel_ExpandList_RepointsCanonicalRawAndLdrRefs_AndUndoRestores` — **real ROM-state** round-trip: plants the canonical pointer + a 3-row table + a **secondary raw pointer** + a Thumb `ldr r0,[pc,#0]` literal-pool word holding the old base; `ExpandList(newCount, undo)` under one undo transaction → **canonical + raw + LDR-literal word all repoint to newBase**; `RunUndo()` restores all three to oldBase **and** restores the wiped old-region bytes
- [x] `ViewModel_ExpandList_CleanRom_NoSecondaryRefs_SucceedsAndCanonicalPointsToNewBase` — `RepointAllReferences` returns 0 = success (no rollback)
- [x] `ViewModel_ExpandList_Signature_TakesUndoData_AndCallsRepointAllReferences` — 2-arg `(uint, Undo.UndoData)` overload; view passes `_undoService.GetActiveUndoData()`; stale "without the LDR-pointer rescan" doc gone
- [x] Updated the 3 existing `ExpandList(` smoke-test callers to the 2-arg signature
- [x] GUI launch on `roms/FE8U.gba` — see GUI Test Report

## GUI Test Report
| Test | Result |
|---|---|
| Map Action Animation editor opens with full surface (Animation Pointer, Padding, Comment, Export/Import, **List Expansion**) | ✅ PASS (screenshot) |
| `ExpandList` compiles + runs through the new `RepointAllReferences` path | ✅ PASS (build + 31 filtered tests) |
| LDR-literal + raw + canonical repoint on expand, and undo restore | ✅ PASS — **real ROM-state unit test** (the authoritative correctness proof; see test plan) |

**Note on the screenshot:** vanilla FE8U/FE8J ship an **empty** map-action-animation table (Read Count: 0), so a *visually populated* expand cannot be demonstrated on the available commercial ROMs — the editor's "List Expansion" affordance is shown, and the actual LDR-literal/raw/canonical repoint + undo behavior is proven by the **real ROM-state** unit test `ViewModel_ExpandList_RepointsCanonicalRawAndLdrRefs_AndUndoRestores` (not a source-only assertion). This mirrors the validation approach for the already-merged `WorldMapImageListExpandTests`.

## Screenshots
**Map Action Animation editor — the affected editor with the List Expansion affordance (LDR-repoint behavior validated by the real ROM-state unit test above):**
Map Action Animation editor

---
Original request: *"set ralph loop to resolve all open GitHub issues in laqieer/FEBuilderGBA"* — 8th issue in the autonomous sweep.

🤖 Generated with [Claude Code](https://claude.com/claude-code) `v2.1.168` · model `claude-opus-4-8[1m]`

### CODE REVIEW

**Review by laqieer — COMMENTED** [review]
No blocking findings. The implementation matches the approved #1025 plan: `ExpandList` captures the old table base, runs `ExpandTableTo`, then calls `DataExpansionCore.RepointAllReferences` as the final ROM-mutating step while threading the active `UndoData`, so the canonical slot, secondary raw pointers, and Thumb LDR literal-pool references are recorded in the same undo transaction.

The key regression test `ViewModel_ExpandList_RepointsCanonicalRawAndLdrRefs_AndUndoRestores` is behavior-level ROM-state coverage, not just a source assertion: it plants canonical + raw + LDR-literal references, verifies they all move to `newBase`, and verifies undo restores the references and wiped old bytes. I also accepted the screenshot/test-report combination here because the affected commercial ROM data is empty; the release-asset screenshot shows the running affected editor and the data-level behavior is covered by the real ROM-state test. The PR body has a GUI Test Report table, all Test plan items are checked, and I found no feature-branch blob/raw URLs.

Non-blocking follow-up: `ImageMapActionAnimationView.axaml.cs` still has an XML-doc sentence saying the flow is "minus the LDR rescan". The code path is correct, so this is not a blocker, but updating that comment would avoid future confusion.

Copilot CLI: 1.0.60
Model: GPT-5.5 (gpt-5.5)

File: FEBuilderGBA.Avalonia/ViewModels/ImageMapActionAnimationViewModel.cs [hunk 1]
```diff
@@ -534,11 +542,22 @@ public string ExpandList(uint newCount)
...
                 return result.Error ?? R._("Table expansion failed.");
 
-            // Refresh the read-config from the new pointer base.
+            // Repoint EVERY raw 32-bit + ARM-Thumb LDR literal-pool reference to
+            // the old base — not just the canonical pointer ExpandTableTo already
+            // moved. RepointAllReferences returning 0 (clean ROM, no secondary
+            // refs) is SUCCESS — do NOT roll back on 0. The canonical slot now
+            // holds the new base and is therefore NOT re-matched (no double-write).
+            // MUST be the LAST ROM-mutating call.
+            DataExpansionCore.RepointAllReferences(rom, oldBase, result.NewBaseAddress, undo);  // #1025: also repoint raw + LDR-literal refs
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/ViewModels/ImageMapActionAnimationViewModel.cs]: Fixed in fb3b28b03 — ExpandList now passes `null` to RepointAllReferences so it relies on the View's ambient UndoService scope instead of opening a second overwrite-style ROM.BeginUndoScope. Empirically verified: the undo round-trip test still restores all three refs (canonical+raw+LDR) and the wiped old bytes.

File: FEBuilderGBA.Avalonia/Views/ImageMapActionAnimationView.axaml.cs [hunk 2]
```diff
@@ -247,7 +247,7 @@ async void ListExpand_Click(object? sender, RoutedEventArgs e)
                 _undoService.Begin("Expand Map Action Animation List");
                 try
                 {
-                    string err = _vm.ExpandList(newCount);
+                    string err = _vm.ExpandList(newCount, _undoService.GetActiveUndoData());
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/Views/ImageMapActionAnimationView.axaml.cs:252]: Fixed in fb3b28b03 — updated the ListExpand_Click XML doc: ExpandList now composes DataExpansionCore.RepointAllReferences (raw 32-bit + ARM-Thumb LDR literal-pool repoint), so the LDR rescan is no longer a gap.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 3]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+        // after, so `p - sig.Length - 4` lands on the pointer (the canonical
+        // slot FindAnimationPointer returns).
+        uint searchStart = 0x1000000;
+        uint pointerAddr = searchStart;
+        uint sigAddr = pointerAddr + 4;
+        canonicalSlot = pointerAddr;
+
+        // Canonical pointer (GBA format = base + 0x08000000).
+        BitConverter.GetBytes(LdrTableBase | 0x08000000u).CopyTo(bytes, pointerAddr);
+        sig.CopyTo(bytes, sigAddr);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 4]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+
+        // Table rows: row 0 = 0 (reserved-null), rows 1..rowCount-1 = valid
+        // pointers (per-row distinct marker), row `rowCount` = 0xFFFFFFFF
+        // terminator so LoadList stops at exactly `rowCount` rows.
+        uint size = ImageMapActionAnimationViewModel.SIZE; // 8
+        BitConverter.GetBytes(0u).CopyTo(bytes, LdrTableBase + 0);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 5]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+        for (int i = 1; i < rowCount; i++)
+        {
+            uint row = LdrTableBase + (uint)i * size;
+            BitConverter.GetBytes((0x00200000u + (uint)i * 0x100u) | 0x08000000u).CopyTo(bytes, row + 0);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 6]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+        }
+        BitConverter.GetBytes(0xFFFFFFFFu).CopyTo(bytes, LdrTableBase + (uint)rowCount * size);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 7]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+
+        // Known 0xFF free region for ExpandTableTo to relocate into.
+        for (int i = 0; i < 0x4000; i++)
+            bytes[LdrFreeRegion + i] = 0xFF;
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.

File: FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs [hunk 8]
```diff
@@ -672,10 +672,262 @@ public void ViewModel_ImportScript_WithoutEntry_ReturnsError()
...
+
+        if (plantSecondaryRefs)
+        {
+            // SECOND raw 32-bit pointer to LdrTableBase.
+            BitConverter.GetBytes(LdrTableBase | 0x08000000u).CopyTo(bytes, LdrRawSlot);
+            // ARM Thumb LDR r0,[pc,#0] (0x4800) + its literal-pool slot.
+            bytes[LdrInstr + 0] = 0x00;
+            bytes[LdrInstr + 1] = 0x48;
+            BitConverter.GetBytes(LdrTableBase | 0x08000000u).CopyTo(bytes, LdrLiteralSlot);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia.Tests/GapSweep/ImageMapActionAnimationParityTests.cs]: These compile fine via the `long` overloads of Array.CopyTo / the byte[] indexer (uint→long is implicit) — CI is green, proving compilation. Cast the buffer indices to (int) in fb3b28b03 for clarity anyway.
